### PR TITLE
CRM-20958 - Track creation+modification times for activities+cases

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------
In developing workflows, reports, and UIs for activities and cases, it is useful to sort and filter based based on when a record was created or when a record was last modified.

Before
----------------------------------------

 * The tables `civicrm_activity` and `civicrm_case` do __not__ have the timestamp columns `created_date` and `modified_date`.
 * The table `civicrm_log` includes *some* records to indicate creation/modification time:
   * It __does__ include records for activities which have been created or updated via BAO. (Notwithstanding the possibility of local-admin pruning.)
   * It __may not__ include records for sample activities created during installation or others created via SQL.
   * It __does not__ include records for cases. (According to the original design philosophy, all changes to cases would be reflected in the activities, so I suppose `civicrm_log` would have been ~~redundant~~ low-priority.)
   * The timestamps stored in `civicrm_log.modified_date` are not entirely reliable -- they are stored as `DATETIME` without any adjustments for timezone. This is a pre-existing and systemic problem with fields being flagged as `DATETIME` instead of `TIMESTAMP`.

After
----------------------------------------

 * The tables `civicrm_activity` and `civicrm_case` __do__ have the timestamp columns `created_date` and `modified_date`.
   * These fields can be __read__ via APIv3/BAO/DAO/SQL. They cannot be directly __modified__ via APIv3/BAO.
   * These fields are automatically __initialized__ and __updated__ whenever one:
     * Creates or updates an `Activity` via APIv3/BAO/DAO/SQL.
     * Creates or updates a `Case` via APIv3/BAO/DAO/SQL.
     * Creates or updates an `Activity` via "Contacts => New Activity", "View Contact => Actions", or "View Contact => Activities". (This stems from SQL triggers, but it's been tested separately.)
     * Creates or updates a `Case` via "Cases => New Case" or "Manage Cases" UI. (This stems from SQL triggers, but it's been tested separately.)
 * The system status check displays __NOTICEs__ if...
   * There are any case or activity records which do not have `created_date` or `modified_date`.
   * Any columns are marked as `DATETIME` instead of a more sensible `TIMESTAMP`. (This addresses both `civicrm_log.modified_date` and the pre-existing discrepancies left over from [CRM-9683](https://issues.civicrm.org/jira/browse/CRM-9683)).
 * Both __NOTICEs__ suggest installation of the new (experimental) extension, [doctorwhen](https://github.com/civicrm/org.civicrm.doctorwhen), to handle migration+cleanup.
 
screen shot 2017-07-31 at 8 45 53 pm

If the admin installs `doctorwhen`, then a new item appears in "Administer => Doctor When":
screen shot 2017-07-31 at 8 59 33 pm

Technical Details
----------------------------------------
 * For initializing/maintaining the fields `{civicrm_activity,civicrm_case}.{created_date,modified_date}`, there are several SQL triggers. The configuration closely parallels `civicrm_contact`:
   * The SQL column definition for `modified_date` specifies `ON UPDATE CURRENT_TIMESTAMP`). (Note: MySQL only allows one column to be flagged with a default value of `CURRENT_TIMESTAMP`.)
   * The services `civi.activity.triggers` and `civi.case.triggers` install SQL triggers which initialize the `created_date`.
   * The services `civi.activity.triggers` and `civi.case.triggers` install SQL triggers which update `modified_date` whenever custom-data changes.
   * The service `civi.case.staticTriggers` installs SQL triggers which update the `civicrm_case.modified_date` whenever a related activity is created, modified, or deleted.
 * The list of SQL columns to consider migrating is tracked in `CRM_Utils_Check_Component_Timestamps`.

Comments
----------------------------------------
 * The JIRA ticket [CRM-20958](https://issues.civicrm.org/jira/browse/CRM-20958) includes a lengthy list of acceptance criteria.
 * This PR description has been updated to reflect issues that were discussed during development.

---

 * [CRM-20958: Data model: Track creation+modification times for activities+cases](https://issues.civicrm.org/jira/browse/CRM-20958)

### CODE REVIEW

File: CRM/Upgrade/Incremental/php/FourSeven.php [hunk 1]
```diff
@@ -421,6 +421,90 @@ public function upgrade_4_7_24($rev) {
...
+    list($minId, $maxId) = CRM_Core_DAO::executeQuery(
+      "SELECT coalesce(min(id),0), coalesce(max(id),0) FROM civicrm_activity"
+    )->getDatabaseResult()->fetchRow();
+    for ($startId = $minId; $startId <= $maxId; $startId += self::BATCH_SIZE) {
+      $endId = $startId + self::BATCH_SIZE - 1;
+      $vars = array(1 => array($startId, 'Int'), 2 => array($endId, 'Int'));
+
+      $title = sprintf('CRM-20958 - Compute civicrm_activity.created_date from civicrm_log (%d => %d)', $startId, $endId);
+      $sql = 'UPDATE civicrm_activity
+       SET created_date = (SELECT MIN(l.modified_date) FROM civicrm_log l WHERE l.entity_table ="civicrm_activity" AND civicrm_activity.id = l.entity_id)
```
Comment (totten) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: One thing I'm not certain about... this code initializes `civicrm_activity.created_date` (etal) using `civicrm_log.modified_date`, but they have different data-types (`TIMESTAMP` vs `DATETIME`, respectively).

@eileenmcnaughton @seamuslee001 To my recollection, `DATETIME` and `TIMESTAMP` handle timezones differently, so wondering if there's some kind of filtering I should do here to adjust for timezone.

However, my gut says that storing `civicrm_log.modified_date` as `DATETIME` is actually a subtle dirtiness/corruption. Ex: [This](https://github.com/civicrm/civicrm-core/blob/4.7.22/CRM/Activity/BAO/Activity.php#L633-L640) suggests that `civicrm_log.modified_date` would be initialized based on `date('YmdHis')` (ie the user's *perceived timestamp* based on active PHP timezone setting) rather than a *canonical timestamp*. But different `civicrm_log` records can be created by different people/sessions/timezones, so the raw data in `civicrm_log` wouldn't be reliably articulated in any one timezone.

This wouldn't really the fault of my issue/patch -- if the theory is true, you'd expect to see idiosyncrasies in the display of this data. For migration purposes, somewhere you'd have to bite the bullet and accept that timestamps would be wrong +/- a few hours?

Comment (totten) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: Confirmed:
 * Open a new browser. Login to `d46` as `admin`. Set your timezone to `America/New_York`. Edit contact 139. Observe "Change Log" has new entry with timestamp `July 26th, 2017 7:29 PM`.
 * Open a new browser. Login to `d46` as `demo`. Set your timezone to `America/Los_Angeles`. Edit contact 139. Observe "Change Log" has new entry with timestamp `July 26th, 2017 4:30 PM`. (In reality, this is 1 minute later.)

But the changelog (as viewed by either user) shows weird data: `admin`'s change actually happened **_1 minute before_** `demo`'s change, but it looks like it happened _**179 minutes after**_.

screen shot 2017-07-26 at 4 34 29 pm

**Comment by seamuslee001** [comment]
Yep so timestamp stores everything in UTC and does conversion to UTC if connection is in another Timezone. Date time is just what ever time is sent to it from the client

**Comment by xurizaemon** [comment]
CRM-9683 seems relevant to all this - same basic problem, over in CiviMail?

**Comment by seamuslee001** [comment]
I thought that CiviMail fields had been converted to timestamp already

**Comment by seamuslee001** [comment]
Maybe that was only on the Fuzion branch of things

**Comment by eileenmcnaughton** [comment]
@seamuslee001 we have converted some fields on new installs but not on existing due to nervousness about the - the fundamental thing I believe it when you convert you should be in the most accurate tz possible

**Comment by seamuslee001** [comment]
@eileenmcnaughton for some reason i thought the CiviMail was already converted, maybe AUG / fuzion thing

**Comment by eileenmcnaughton** [comment]
@seamuslee001 yes, converted on all Fuzion customers & any site that wishes to convert can do so with a simple ALTER TABLE mysql statement. Also converted for new installs - but we have not figured out what to do with existing ones (mostly because the timestamp will be relative to the tz used when converting & there is some room for debate there)

**Comment by eileenmcnaughton** [comment]
(there is a case to be made for providing a change over routine that people can run through UI or cli depending which is more appropriate)

**Comment by totten** [comment]
> (there is a case to be made for providing a change over routine that people can run through UI or cli depending which is more appropriate)

Yeah, this is my main question at this point. 

On one hand... this PR seems to be working for me. The schema is created; the data migrates according to the evil scenario described in the commit-notes; the API seems to return and sort the fields; the fields for cases+activities are updated when using "Manage Case" UI; there's new test coverage; and the existing tests are now passing.

On the other hand... it might be a safer phase-in process if we only add the new columns -- and put the other stuff (the triggers and migration) behind some kind of opt-in.

**Comment by eileenmcnaughton** [comment]
I think the triggers should be uncontroversial - since they are also in the contact schema. The question is more the migration. On balance I probably would, since that was done for the civicrm_contact table. But if you were going to leave something off I would limit it to the migration

**Comment by totten** [comment]
@eileenmcnaughton @xurizaemon @seamuslee001  I've moved the migration logic to an extension https://github.com/civicrm/org.civicrm.doctorwhen and put a status check in `CRM_Utils_Check_Component_Timestamps`. Additionally, since CRM-9683 sorta left the transition from `DATETIME` to `TIMESTAMP` in a netherrealm, and since `doctorwhen` is labelled "experimental", it also  includes checkboxes for migrating those columns.

**Comment by eileenmcnaughton** [comment]
cool  - I wonder how people would find out about that if they wanted to switch

**Comment by totten** [comment]
CC @demeritcowboy - you might find this PR interesting.

**Comment by seamuslee001** [comment]
makes sense to me

**Comment by demeritcowboy** [comment]
Thanks for the notice. Hmm, DoctorWhen - Timezones and Relative Dates in Civi?

Don't think it makes a difference here but noting the original use of civicrm_log wasn't entirely redundant - for example to see an activity's original creation date you still needed to look in civicrm_log.

It also probably doesn't make a difference but since the opening paragraph mentions it as the rationale, noting that at least in my experience Activity Date (activity_date_time) is usually the field used for sorting/reporting, and it allows users to have some control over how things sort without messing up the "system" dates like creation date. The system dates are usually just used during investigation. Cases and some of their activities are often allocated to certain reporting periods via the activity date, similar to financial reporting. But none of that prevents making changes for create/modify.

File: tests/phpunit/api/v3/ActivityTest.php [hunk 2]
```diff
@@ -447,6 +447,11 @@ public function testActivityCreateExample() {
     require_once 'api/v3/examples/Activity/Create.php';
     $result = activity_create_example();
     $expectedResult = activity_create_expectedresult();
+    // Mask timestamp fields. Too variable.
```
Comment (davialexandre) [inline tests/phpunit/api/v3/ActivityTest.php]: Sorry, but I'm afraid I don't understand this comment. Is is because the timestamps will change every time you run the test?

File: tests/phpunit/api/v3/CaseTest.php [hunk 3]
```diff
@@ -165,6 +165,9 @@ public function testCaseUpdate() {
 
     // Verify that updated case is exactly equal to the original with new subject.
     $result = $this->callAPISuccessGetSingle('Case', array('case_id' => $id));
+    $this->assertTrue($result['modified_date'] >= $case['modified_date']);
+    unset($result['modified_date']);
+    unset($case['modified_date']);
```
Comment (davialexandre) [inline tests/phpunit/api/v3/CaseTest.php]: What about leaving a comment here to explain why the unsets are necessary?

File: tests/phpunit/api/v3/CaseTest.php [hunk 4]
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
...
+      'case_id' => $case_created['id'],
+      'options' => array(
+        'limit' => 1,
+      ),
+    ));
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_1['created_date']);
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_1['modified_date']);
+    $this->assertApproxEquals(strtotime($activity_1['created_date']), strtotime($activity_1['created_date']), 2);
+
+    usleep(1.5 * 1000000);
```
Comment (davialexandre) [inline tests/phpunit/api/v3/CaseTest.php]: Why is this necessary?

File: tests/phpunit/api/v3/CaseTest.php [hunk 5]
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
...
+  public function testTimestamps() {
+    $params = $this->_params;
+    $case_created = $this->callAPISuccess('case', 'create', $params);
+
+    $case_1 = $this->callAPISuccess('Case', 'getsingle', array(
+      'id' => $case_created['id'],
+    ));
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $case_1['created_date']);
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $case_1['modified_date']);
+    $this->assertApproxEquals(strtotime($case_1['created_date']), strtotime($case_1['created_date']), 2);
```
Comment (davialexandre) [inline tests/phpunit/api/v3/CaseTest.php]: You're comparing the created_date with itself here. Is this correct?

File: tests/phpunit/api/v3/CaseTest.php [hunk 6]
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
...
+
+    $activity_1 = $this->callAPISuccess('activity', 'getsingle', array(
+      'case_id' => $case_created['id'],
+      'options' => array(
+        'limit' => 1,
+      ),
+    ));
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_1['created_date']);
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_1['modified_date']);
+    $this->assertApproxEquals(strtotime($activity_1['created_date']), strtotime($activity_1['created_date']), 2);
```
Comment (davialexandre) [inline tests/phpunit/api/v3/CaseTest.php]: Here too

File: tests/phpunit/api/v3/CaseTest.php [hunk 7]
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
...
+    $activity_2 = $this->callAPISuccess('activity', 'getsingle', array(
+      'id' => $activity_1['id'],
+    ));
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_2['created_date']);
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_2['modified_date']);
+    $this->assertNotEquals($activity_2['created_date'], $activity_2['modified_date']);
+
+    $this->assertEquals($activity_1['created_date'], $activity_2['created_date']);
+    $this->assertNotEquals($activity_1['modified_date'], $activity_2['modified_date']);
+    $this->assertTrue(strtotime($activity_1['modified_date']) < strtotime($activity_2['modified_date']),
```
Comment (davialexandre) [inline tests/phpunit/api/v3/CaseTest.php]: What about using `assertLessThan()` instead of `assertTrue()`?

File: tests/phpunit/api/v3/CaseTest.php [hunk 8]
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
...
+      sprintf("Original modification time (%s) should predate later modification time (%s)", $activity_1['modified_date'], $activity_2['modified_date']));
+
+    $case_2 = $this->callAPISuccess('Case', 'getsingle', array(
+      'id' => $case_created['id'],
+    ));
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $case_2['created_date']);
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $case_2['modified_date']);
+    $this->assertEquals($case_1['created_date'], $case_2['created_date']);
+    $this->assertNotEquals($case_2['created_date'], $case_2['modified_date']);
+  }
```
Comment (davialexandre) [inline tests/phpunit/api/v3/CaseTest.php]: I must confess it was a bit hard for me to read this test and understand what was going on. I feel it would have been easier if it was split into smaller tests, each one having a descriptive name of what is being tested. Examples:
- `testCreatedDateShouldBeTheSameAsModifiedDateWhenACaseIsCreated`
- `testCreatedDatedIsNotUpdatedDuringWhenACaseIsUpdated`
- `testModifiedDateShouldBeUpdatedWhenACaseIsUpdated`
- `testCreatedDateAndModifiedDateCannotBeManuallyChanged`
- `testDateFormatOfCreatedDateAndModifiedDateShouldBeCorrect`

Note that the examples have some scenarios that were not even covered by tests. In addition to that, do you think it would make sense to also have tests for the BAO (that is, check that the same rules will work when you do something like `Case::create()`?

File: CRM/Upgrade/Incremental/php/FourSeven.php [hunk 9]
```diff
@@ -1197,4 +1217,24 @@ protected function checkImageUploadDir() {
...
+   * @param string $sql
+   *   An SQL template. May include tokens `%1`, `%2`, etc.
+   * @param array $vars
+   *   List of SQL parameters, as used by executeQuery().
+   *
+   * @return bool
+   *
+   * @see CRM_Core_DAO::executeQuery
+   */
+  public static function task_executeQuery($ctx, $sql, $vars) {
```
Comment (davialexandre) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: Maybe I missed it, but it doesn't look like you're using this anywhere. Are you?

**Review by davialexandre — CHANGES_REQUESTED** [review]
hi @totten, left some inline comments

File: CRM/Utils/Check/Component/Timestamps.php [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+ * @copyright CiviCRM LLC (c) 2004-2017
+ */
+class CRM_Utils_Check_Component_Timestamps extends CRM_Utils_Check_Component {
+
+  const DOCTOR_WHEN = 'https://github.com/civicrm/org.civicrm.doctorwhen';
+
+  /**
+   * Check that the timestamp columns are populated. (CRM-20958)
+   *
+   * @return array<CRM_Utils_Check_Message>
```
Comment (totten) [inline CRM/Utils/Check/Component/Timestamps.php]: It means "an array in which the elements are instances of class `CRM_Utils_Check_Message`".

Personally, I go a bit bonkers reading `@return array` because it doesn't really sanything.

AFAIK, `php.net` hasn't blessed a notation for array typing, but there a few references for comparison:
 * It's similar to Java's notation for generics (eg `Vector<Message>`, `HashMap<String,Message>`).
 * It matches [Hack's array notation](https://docs.hhvm.com/hack/collections/introduction). (Hack is PHP derivative with stronger typing.)
 * PHP-FIG has  [a draft spec](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#abnf) for docblocks. It mentions both C-style arrays (`Message[]`) as well Java-style generics (`Collection<Message>`).
 * In my local copy of PHPStorm, it provides drilldown for C-style arrays (`Message[]`) but not Java-style generics.
 * I'm not particularly consistent. Sometimes it seems meaningful to convey even more information, so I use really verbose descriptor like:

```
  @return array
     Array(string $msgId => CRM_Utils_Check_Message $msg).
```

I guess standardizing/documenting/cleaning-up would be good, but that should be probably be a separate project.

File: CRM/Utils/Check/Component/Timestamps.php [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  /**
+   * Check that various columns are TIMESTAMP and not DATETIME. (CRM-9683, etal)
+   *
+   * @return array
+   */
+  public function checkSchema() {
+    $problems = array();
+    foreach (self::getConvertedTimestamps() as $tgt) {
```
Comment (totten) [inline CRM/Utils/Check/Component/Timestamps.php]: I'll rename it to `$target`.

File: Civi/Core/SqlTriggers/StaticTriggers.php [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+namespace Civi\Core\SqlTriggers;
+
+/**
+ * Build a set of SQL simple, literal SQL triggers.
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2017
+ */
+class StaticTriggers {
```
Comment (totten) [inline Civi/Core/SqlTriggers/StaticTriggers.php]: Re: `SqlTriggers` namespace -- Agree that should be singular. Pretty much all namespaces can be equally described as singular or plural, and it's nice to have consistent style.

Re: `StaticTriggers` class -- I don't think the singular/plural issue as a matter of style here -- it's really a semantic difference. `SqlTrigger` would just be one object managing one trigger. But if one object manages multiple triggers, then you'd say `SqlTriggers` or `SqlTriggerSet` or `SqlTriggerList` or `SqlTriggerArray` or `SqlTriggerGroup`. Personally, I prefer for the simplest construction that's accurate.

In this case, this instance of `StaticTriggers` contains three triggers that serve a common goal. (The goal is to update `civicrm_case.modified_date` whenever a related activity is inserted, updated, or deleted.)

File: Civi/Core/SqlTriggers/StaticTriggers.php [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+  /**
+   * @param \Civi\Core\Event\GenericHookEvent $e
+   * @see \CRM_Utils_Hook::triggerInfo
+   */
+  public function onTriggerInfo($e) {
+    $this->alterTriggerInfo($e->info, $e->tableName);
+  }
+
+  /**
+   * Get a list of triggers for the contact table.
```
Comment (totten) [inline Civi/Core/SqlTriggers/StaticTriggers.php]: Yup, will fix.

File: Civi/Core/SqlTriggers/StaticTriggers.php [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+   *
+   * @see \CRM_Utils_Hook::triggerInfo
+   * @see \CRM_Core_DAO::triggerRebuild
+   * @see http://issues.civicrm.org/jira/browse/CRM-10554
+   *
+   * @param $info
```
Comment (totten) [inline Civi/Core/SqlTriggers/StaticTriggers.php]: It's a good question, but the answer is to drill-down on `@see \CRM_Utils_Hook::triggerInfo`. I don't want to reproduce the hook spec here. Will put another comment directing folks there.

File: Civi/Core/SqlTriggers/StaticTriggers.php [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+   * @param null $tableFilter
```
Comment (totten) [inline Civi/Core/SqlTriggers/StaticTriggers.php]: It's a good question, but the answer is to drill-down on `@see \CRM_Utils_Hook::triggerInfo`. I don't want to reproduce the hook spec here. Will put another comment directing folks there.

File: tests/phpunit/api/v3/CaseTest.php (hunk 4 above)
Comment (totten) [inline tests/phpunit/api/v3/CaseTest.php]: It ensures that subsequent modifications will have a new timestamp (greater than the `created_date` or previous modifications).

File: CRM/Upgrade/Incremental/php/FourSeven.php (hunk 9 above)
Comment (totten) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: Valid point. It was a helper to support the migration task, but then the "migration task" migrated itself off to `doctorwhen`, so we don't really need this straggler.

OTOH, it is a handy straggler...

File: tests/phpunit/api/v3/CaseTest.php (hunk 5 above)
Comment (totten) [inline tests/phpunit/api/v3/CaseTest.php]: Thanks, I'll fix that.

File: tests/phpunit/api/v3/CaseTest.php (hunk 3 above)
Comment (totten) [inline tests/phpunit/api/v3/CaseTest.php]: Done

File: tests/phpunit/api/v3/ActivityTest.php (hunk 2 above)
Comment (totten) [inline tests/phpunit/api/v3/ActivityTest.php]: Yup, pretty much. I'll rephrase it.

File: Civi/Core/SqlTriggers/TimestampTriggers.php [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+    if (\CRM_Core_Config::isUpgradeMode()) {
+      if (!\CRM_Core_DAO::checkFieldExists($this->getTableName(),
+        $this->getCreatedDate())
+      ) {
+        return;
+      }
+      //      $currentVer = CRM_Core_BAO_Domain::version(TRUE);
+      //      if (version_compare($currentVer, '4.3.alpha1') < 0) {
+      //        return;
+      //      }
```
Comment (totten) [inline Civi/Core/SqlTriggers/TimestampTriggers.php]: Analytically, you're right, the commented code is superfluous and may confuse future readers. Emotionally, based on some of the history, I wanted to keep the example handy for a month. But it's not that important. Reworking the comments.

File: CRM/Utils/Check/Component/Timestamps.php (hunk 10 above)
Comment (davialexandre) [inline CRM/Utils/Check/Component/Timestamps.php]: I believe that most projects use the notation defined by http://phpdoc.org, which uses the `CRM_Utils_Check_Message[]`. Given that the PHP-FIG draft is a derivative of that phpdoc and that most projects already use the C-style arrays, I think safe enough to assume this is the "right" way to do it.

Sometimes I prefer a more verbose description too

File: Civi/Core/SqlTriggers/StaticTriggers.php (hunk 12 above)
Comment (davialexandre) [inline Civi/Core/SqlTriggers/StaticTriggers.php]: I can't say that I agree, but I don't have a better name to suggest. Since you think it's good enough, I won't argue :)

File: tests/phpunit/api/v3/CaseTest.php [hunk 17]
```diff
@@ -163,8 +163,13 @@ public function testCaseUpdate() {
     $params['subject'] = $case['subject'] = 'Something Else';
     $this->callAPISuccess('case', 'create', $params);
 
-    // Verify that updated case is exactly equal to the original with new subject.
+    // Verify that updated case is equal to the original with new subject.
     $result = $this->callAPISuccessGetSingle('Case', array('case_id' => $id));
+    // Modification dates are likely to differ by 0-2 sec. Check manually.
+    $this->assertTrue($result['modified_date'] >= $case['modified_date']);
```
Comment (davialexandre) [inline tests/phpunit/api/v3/CaseTest.php]: I saw you accepted my suggestion to use `assertLessThan()` so maybe can you use `assertGreaterThanOrEqual` here, especially because there's no custom message for this assertion

**Comment by totten** [comment]
@agh1 @MegaphoneJon @Stoob  - I think you guys are pretty thoughtful about messaging. Could you comment on how well these screens present the matter of transitioning/cleaning-up dates/times? See the screenshots and [hyperlink](https://github.com/civicrm/org.civicrm.doctorwhen) from the PR description.

In years past, we might have tried to address the change/cleanup of time data with an automated upgrade script, and then there'd be fall-out when the script doesn't anticipate some issue... which is fairly likely because no one has access to the kind of context/tools/datasets needed to ensure a change works everywhere. In theory, the RC period helps, but I'm feeling a little untrusting there. So this PR tries to break it into smaller steps -- eg make the smallest change possible, show a notification, say "Don't panic", and then link to experimental docs/tools for the rest.

The in-app messages will probably set the tone+reactions. Most of the folks on this ticket have already formed some perspective (due historical issues) -- so I wanted to ping you guys for a fresh perspective on the messaging/transition-path.

**Comment by totten** [comment]
Following up on my questions about messaging... I escalated to `civicrm-dev` and moved half of the messaging to separate/smaller JIRA issue and PR (CRM-21079, #10874), which reduced the scope of the current PR.

Also, just rebased to handle a merge-conflict.
